jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java

```diff
@@ -59,46 +77,120 @@
-		} catch (Exception e) {
-			logger.warn("Execution of time failed due to {}", e.getMessage(), e);
-			throw new RuntimeException(e);
+        try {
+            transaction(this::executeTimerJobInstance, timerJobInstance);
+        } catch (Exception e) {
+            recoverTimerJobInstance(timerJob, e);
 		}
 	}
 
+    private void executeTimerJobInstance(TimerJobInstance timerJobInstance) throws Exception {
+        try {
+            ((Callable<?>) timerJobInstance).call();
+        } catch (Exception e) {
+            logger.warn("Execution of time failed due to {}", e.getMessage(), e);
+            throw e;
+        }
+    }
+
+    private void recoverTimerJobInstance(EjbTimerJob ejbTimerJob, Exception e) {
+        // if we have next date fired means that it would have been reescheduled already by DefaultTimerJobInstance
+        if (ejbTimerJob.getTimerJobInstance().getTrigger().hasNextFireTime() != null) {
+            logger.warn("Execution of time failed Interval Trigger failed {}", ejbTimerJob.getTimerJobInstance());
+            return;
+        }
+
+        // if there is not next date to be fired, we need to apply policy otherwise will be lost
+        switch (TIMER_RETRY_POLICY) {
+            case RETRY:
+                logger.warn("Execution of time failed. The timer will be retried {}", ejbTimerJob.getTimerJobInstance());
+                Transaction<TimerJobInstance> operation = (instance) -> {
+                    ZonedDateTime nextRetry = ZonedDateTime.now().plus(TIMER_RETRY_INTERVAL, ChronoUnit.MILLIS);
+                    EjbTimerJobRetry info = null;
+                    if(ejbTimerJob instanceof EjbTimerJobRetry) {
+                        info = ((EjbTimerJobRetry) ejbTimerJob).next();
+                    } else {
+                        info =  new EjbTimerJobRetry(instance);
+                    }
+                    if (TIMER_RETRY_LIMIT > 0 && info.getRetry() > TIMER_RETRY_LIMIT) {
+                        logger.warn("The timer {} reached retry limit {}. It won't be retried again", instance, TIMER_RETRY_LIMIT);
+                        return;
+                    }
+                    TimerConfig config = new TimerConfig(info, true);
+                    timerService.createSingleActionTimer(Date.from(nextRetry.toInstant()), config);
+                };
+                try {
+                    transaction(operation, ejbTimerJob.getTimerJobInstance());
+                } catch (Exception e1) {
+                    logger.error("Failed to executed timer recovery {}", e1.getMessage(), e1);
+                }
+                break;
+            case PLATFORM:
+                logger.warn("Execution of time failed. Application server policy applied {}", ejbTimerJob.getTimerJobInstance());
+                throw new RuntimeException(e);
+        }
+    }
+
+    @FunctionalInterface
+    private interface Transaction<I> {
+
+        void doWork(I item) throws Exception;
+    }
```

Comment:
Do we need another functional interface? What about using a simple Consumer here?

Reply:
AFAIF consumer does not allow to throw an exception. I didn't see any functional interface allowing that.

Reply:
That's true, didn't notice that, sorry.